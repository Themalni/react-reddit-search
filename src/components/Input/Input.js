import styled from 'styled-components'

const InputText = styled.input`
  border: 0;
  width: 400px;
  margin-right: 0.5em;
  padding: 1em;
  border-radius: 4px;
  font-size: 1em;
`

export { InputText }