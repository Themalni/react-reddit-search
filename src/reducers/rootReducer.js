import { combineReducers } from 'redux';
import { postsReducer } from './postsReducer';
import searchReducer from './searchReducer';

export default combineReducers({
  posts: postsReducer,
  search: searchReducer
})