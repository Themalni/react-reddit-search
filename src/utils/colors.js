export const colors = {
  white: '#fff',
  graphite: '#666',
  grey: '#8a96a3',
  yellowOrange: '#ff9933',
  orange: '#ed5f55',
  orangeDark: '#d8574e',
  blue: '#3a81b7',
  blueLight: '#64a0ce'
};

// d02216