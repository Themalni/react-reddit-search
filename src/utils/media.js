export const media = {
  phone: '@media (min-width: 320px)',
  tablet: '@media (min-width: 728px)',
  laptop: '@media (min-width: 1024px)',
  desktop: '@media (min-width: 1280px)'
}