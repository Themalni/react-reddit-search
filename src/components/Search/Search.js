import React, { Component } from 'react';
import { Container } from '../Container/Container';
import Form from '../Form/Form';
import { InputText } from '../Input/Input';
import { Button } from '../Button/Button';
import { connect } from 'react-redux';
import { searchPosts } from '../../actions/searchActions';
import { fetchPosts, isLoadingStateChange, postsNotFoundChangeState } from '../../actions/postsActions';
import PropTypes from 'prop-types';

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      search: ''
    }
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
    this.props.postsNotFoundChangeState(false);
  }

  onSubmit = (e) => {
    e.preventDefault();

    const searchTerm = {
      search: this.state.search
    }
    this.props.searchPosts(searchTerm)
    this.props.fetchPosts()
    this.props.isLoadingStateChange(true)
  }

  render() {
    return (
      <Container>
        <Form onSubmit={this.onSubmit}>
          <InputText type="text" placeholder="Enter search term..." name="search" onChange={this.onChange} value={this.state.search} required />
          <Button>Search</Button>
        </Form>
      </Container>
    )
  }
}

Search.propTypes = {
  searchPosts: PropTypes.func.isRequired
}

const mapStateToProps = (state) => {
  return {
    searchTerm: state.search.searchTerm,
    posts: state.posts.items,
    postsNotFound: state.posts.postsNotFound
  }
}

export default connect(mapStateToProps, { searchPosts, fetchPosts, isLoadingStateChange, postsNotFoundChangeState })(Search)