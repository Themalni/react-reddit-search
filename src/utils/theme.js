import { colors } from './colors';

export const theme = {
  colors,
  font: {
    thin: 300,
    regular: 400,
    bold: 700,
  },
  media: {
    phone: '@media (min-width: 320px)',
    tablet: '@media (min-width: 728px)',
    laptop: '@media (min-width: 1024px)',
    desktop: '@media (min-width: 1280px)'
  }
}