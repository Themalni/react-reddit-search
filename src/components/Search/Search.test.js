import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import rootReducer from '../../reducers/rootReducer';
import { theme } from '../../utils/theme';
import Search from './Search';
import { InputText } from '../Input/Input';
import { ThemeProvider } from 'styled-components';

afterEach(() => cleanup())

function renderWithRedux(component,
  { initState, store = createStore(rootReducer, initState, applyMiddleware(thunk)) } = {}) {
  return {
    ...render(<Provider store={store}><ThemeProvider theme={theme}>{component}</ThemeProvider></Provider>),
    store
  }
}

describe('Search', () => {

  it("renders correctly", () => {
    const { container } = renderWithRedux(<Search></Search>);
    expect(container.firstChild).toMatchSnapshot();
  })

  it("calls handleSubmit when Search button is clicked ", () => {
    const handleSubmitMock = jest.fn();

    const { getByText } = renderWithRedux(
      <Search onSubmit={handleSubmitMock()}></Search>
    )

    fireEvent.click(getByText(/Search/i))

    expect(handleSubmitMock).toHaveBeenCalledTimes(1)
  })

  it("calls onChange when input value changes", () => {
    const onChangeMock = jest.fn();
    const searchTerm = "books";

    const { getByPlaceholderText } = renderWithRedux(
      <Search>
        <InputText type="text" placeholder="Enter search term..." onChange={onChangeMock()} />
      </Search>
    )
    const input = getByPlaceholderText("Enter search term...");

    fireEvent.change(input, {
      target: { value: searchTerm },
    })

    expect(onChangeMock).toHaveBeenCalledTimes(1)
    expect(input.value).toBe(searchTerm)
  })

})