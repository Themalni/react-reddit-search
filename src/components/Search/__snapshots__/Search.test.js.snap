// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  class="sc-bdVaJa bZcouC"
>
  <form
    class="sc-bxivhb gkJYwJ"
  >
    <input
      class="sc-ifAKCX gcaKdI"
      name="search"
      placeholder="Enter search term..."
      required=""
      type="text"
      value=""
    />
    <button
      class="sc-EHOje cxfMfx"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
